frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx

```diff
@@ -6,18 +6,113 @@ import React from 'react'
+import { AbsenceType } from 'lib-common/generated/enums'
+import RoundIcon from 'lib-components/atoms/RoundIcon'
+import { faThermometer } from 'lib-icons'
+import { FixedSpaceRow } from 'lib-components/layout/flex-helpers'
+import {
+  getTimesOnWeekday,
+  isIrregular,
+  isRegular,
+  isVariableTime
+} from 'lib-common/api-types/child/common'
 
 interface Props {
   operationalDays: OperationalDay[]
   reservations: ChildReservations[]
 }
 
+function renderAbsence(type: AbsenceType) {
+  if (type === 'SICKLEAVE')
+    return (
+      <AbsenceCell>
+        <FixedSpaceRow spacing="xs" alignItems="center">
+          <RoundIcon
+            content={faThermometer}
+            color={colors.accents.violet}
+            size="m"
+          />
+          <div>Sairaus</div>
+        </FixedSpaceRow>
+      </AbsenceCell>
+    )
+
+  return (
+    <AbsenceCell>
+      <FixedSpaceRow spacing="xs" alignItems="center">
+        <RoundIcon content="–" color={colors.primary} size="m" />
+        <div>Muu syy</div>
+      </FixedSpaceRow>
+    </AbsenceCell>
+  )
+}
+
+function renderChildDay(
+  day: OperationalDay,
+  childReservations: ChildReservations
+) {
+  const dailyData = childReservations.dailyData.find((d) =>
+    d.date.isEqual(day.date)
+  )
+
+  if (!dailyData) return null
+
+  if (day.isHoliday && !dailyData.reservation && !dailyData.attendance)
+    return null
+
+  if (dailyData.absence && !dailyData.attendance)
+    return renderAbsence(dailyData.absence.type)
+
+  const serviceTimes = childReservations.child.dailyServiceTimes
+  const serviceTimesAvailable =
+    serviceTimes != null && !isVariableTime(serviceTimes)
+  const serviceTimeOfDay =
+    serviceTimes === null || isVariableTime(serviceTimes)
+      ? null
+      : isRegular(serviceTimes)
+      ? serviceTimes.regularTimes
+      : isIrregular(serviceTimes)
+      ? getTimesOnWeekday(serviceTimes, day.date.getIsoDayOfWeek())
+      : null
+
+  return (
+    <DateCell>
+      <AttendanceTimesRow>
+        <Time>{dailyData.attendance?.startTime ?? '–'}</Time>
+        <Time>{dailyData.attendance?.endTime ?? '–'}</Time>
+      </AttendanceTimesRow>
+      <Gap size="xxs" />
+      <ReservationTimesRow>
+        {dailyData.reservation ? (
+          <>
+            <ReservationTime>
+              {dailyData.reservation?.startTime ?? '–'}
+            </ReservationTime>
+            <ReservationTime>
+              {dailyData.reservation?.endTime ?? '–'}
+            </ReservationTime>
+          </>
+        ) : serviceTimesAvailable && serviceTimeOfDay ? (
+          <>
+            <ReservationTime>{serviceTimeOfDay.start}*</ReservationTime>
+            <ReservationTime>{serviceTimeOfDay.end}*</ReservationTime>
+          </>
+        ) : serviceTimesAvailable && serviceTimeOfDay === null ? (
+          <ReservationTime>Vapaapäivä</ReservationTime>
+        ) : (
+          <ReservationTime>Ei varausta</ReservationTime>
+        )}
```

Comment:
It's pretty hard to follow what's happening here.

Reply:
added comments, not sure how to make the code itself more readable